### ISSUE TITLE
Adjust to latest gyroflow file format

### ISSUE THREAD

**Issue by AdrianEddy** [EmberLightVFX/GyroFlow-to-CSV#6]
Hi, 

In latest dev build, and soon to be v1.5.1, I've changed the format of the extended data a bit. Instead of using `bincode` I now use standard `CBOR`, which should be easier to parse

If `version` == 3, then `gyro_data = cbor2.decoder.loads(raw)`

This was needed because we now have additional data from certain files, like per-frame offset, per-frame exposure time and hopefully soon per-frame lens profile information. Adding all these additional information using `bincode` would be pretty hard to parse in external tools, so I decided to switch to CBOR, which has parsers in all languages

`raw_imu` and `quaternions` fields were removed from `gyro_source`. Instead, there's now `file_metadata` there which contains:
```rust
pub struct FileMetadata {
    pub imu_orientation:     Option<String>,
    pub raw_imu:             Vec<telemetry_parser::util::IMUData>, 
    pub quaternions:         BTreeMap<i64, Quaternion<f64>>,
    pub gravity_vectors:     Option<BTreeMap<i64, Vector3<f64>>>,
    pub image_orientations:  Option<BTreeMap<i64, Quaternion<f64>>>,
    pub detected_source:     Option<String>,
    pub frame_readout_time:  Option<f64>,
    pub frame_rate:          Option<f64>,
    pub camera_identifier:   Option<CameraIdentifier>,
    pub lens_profile:        Option<serde_json::Value>,
    pub lens_positions:      Option<BTreeMap<i64, f64>>,
    pub has_accurate_timestamps: bool,
    pub additional_data:     serde_json::Value,
    pub per_frame_time_offsets: Vec<f64>,
    pub per_frame_data:      Vec<serde_json::Value>,
}

// definition of telemetry_parser::util::IMUData
pub struct IMUData {
    pub timestamp_ms: f64,
    pub gyro: Option<[f64; 3]>,
    pub accl: Option<[f64; 3]>,
    pub magn: Option<[f64; 3]>
}

// definition of CameraIdentifier
pub struct CameraIdentifier {
    pub brand: String,
    pub model: String,
    pub lens_model: String,
    pub lens_info: String,
    pub focal_length: Option<f64>,
    pub camera_setting: String,
    pub fps: usize,
    pub video_width: usize,
    pub video_height: usize,
    pub additional: String,
    pub identifier: String
}
```

Additionally, I added:
- `synced_imu_timestamps` - array of timestamps for every IMU sample, with added interpolated offsets
- `synced_imu_timestamps_with_per_frame_offset` - same as above + per frame time offset, which involves exposure time and custom VSYNC offset from the file metadata - this is the final offset. The timestamps represent middle of the frame, ie. in case of 59.94 video, 0 is the middle row of first frame, 16.68 middle row of the second frame etc.
- `adaptive_zoom_fovs` - per frame smoothed FOV for adaptive zoom

So instead of interpolating the offsets on your side in python, you can just use `imu.timestamp = synced_imu_timestamps_with_per_frame_offset[imu_index]`

```python
file_metadata = cbor2.decoder.loads(zlib.decompress(decode(data['gyro_source']['file_metadata'])))
synced_imu_timestamps_with_per_frame_offset = cbor2.decoder.loads(zlib.decompress(decode(data['gyro_source']['synced_imu_timestamps_with_per_frame_offset'])))

**Comment by AdrianEddy** [comment]
[samples.zip](https://github.com/EmberLightVFX/Gyroflow-to-CSV/files/11646627/samples.zip)

**Comment by AdrianEddy** [comment]
[GyroFlowToCSV.zip](https://github.com/EmberLightVFX/Gyroflow-to-CSV/files/11949546/GyroFlowToCSV.zip)

updated, but not tested and only works with the new version, will need to be updated with `if`s that check version to work with both new and old

**Comment by AndrewHazelden** [comment]
Hi @AdrianEddy,

> updated, but not tested and only works with the new version, will need to be updated with `if`s that check version to work with both new and old

I finally tested out the revised GyroflowToCSV script you posted. To get the script to work successfully with Python v3.10.9 on my laptop I had to adjust the cbor2 function to use the `cbor2._decoder.loads()` instead of `cbor2.decoder.loads()`.

Here is the modified python script should anyone else want to explore it:

[GyroFlowToCSV.zip](https://github.com/EmberLightVFX/Gyroflow-to-CSV/files/13060880/GyroFlowToCSV.zip)

**Comment by AdrianEddy** [comment]
In latest dev build I added an option to export camera data directly to CSV, JSON, USD or JSX (for AE) from Gyroflow. Please test if it works as expected for you.

I also have a question - what's the use case for this data? Since this only exports camera rotations, and not translations, it doesn't replace 3D camera tracker. Is there a way to use rotations and then track translations only?

https://gyroflow.xyz/devbuild/ Motion Data -> Export -> Export camera data

**Comment by AndrewHazelden** [comment]
Thanks for adding the native data export options to GyroFlow. This will make things far more accessible for end users!

**Comment by AndrewHazelden** [comment]
My primary interest in the GyroFlow CSV/JSON/USD data export options is to use the motion data to help with spherical video rotational stabilization for 180VR/360VR post-production workflows.

**Comment by MiloszVR** [comment]
Hi! This is exactly what I'm trying to achieve - use GoPro gyro data to stabilise 360VR shots - but am facing issues:

I successfully installed _Vonk_ and the _Gyroflow CSV macro_ inside Resolve/Fusion Page.
I export a csv file from _Gyroflow_ with the stabilised Euler data (not using _Gyroflow to CSV_ script). I have to delete the first column before I import it into Fusion to have a formatting matching the one demonstrated in the example: VID_20220615_110533_00_001.csv.
That data is then inside Fusion Page successfully applied to the 360 footage but it does not stabilise at all just changing the sphere position in an inconsistent way. I played with the axis orientations and the order of XYZ data but without success.
It is worth mentioning that I’m using an offset in _vNumberAdd1_ Node as I didn’t stitched the whole original GoPro clip. Oh, I’m using a 360 rig with 12 gopros and extracting gyro data from one of them.

I’m using Gyroflow 1.6.1 and the highest Resolve Studio 19.

I feel stuck and not sure what to try next.

Thank you for your help,
Milosz

**Comment by EmberLightVFX** [comment]
In the latest GuroFlow you can export the camera to a fbx or usd (I don't remember what ones) and import it into Fusion. Way easier than this workflow 😄
That's one reason I haven't updated it in a while.

**Comment by xphoniex** [comment]
Hey @AdrianEddy , does a `.gyroflow` file contain focus distance timeseries? Included example [file](https://github.com/EmberLightVFX/Gyroflow-to-CSV/blob/main/example/VID_20220615_110533_00_001.gyroflow) in this repo has these fields:
```
title
version
app_version
videofile
calibration_data
date
background_color
background_mode
background_margin
background_margin_feather
video_info
stabilization
gyro_source
output
synchronization
offsets
trim_start
trim_end
```
which seem to be the same fields as [gyroflow/gyroflow/src/core/lib.rs](https://github.com/gyroflow/gyroflow/blob/81d2f27bea5edb8d47bf91c2dd8a2d3b1108dea7/src/core/lib.rs), but I can't seem to find anything related to lens data or focus distance. Can you help?

**Comment by AdrianEddy** [comment]
Yes, in gyro_source->file_metadata in CBOR compressed with ZLIB and encoded with base91
https://github.com/gyroflow/gyroflow/blob/master/src/core/lib.rs#L1259-L1261
https://github.com/gyroflow/gyroflow/blob/master/src/core/gyro_source/file_metadata.rs
